devito/types/dense.py

```diff
@@ -792,17 +792,21 @@ def _halo_exchange(self):
                 # Gather send data
                 data = self._data_in_region(OWNED, d, i)
                 sendbuf = np.ascontiguousarray(data)
+                if self.dtype == np.float16:
+                    sendbuf = sendbuf.view(np.uint16)
```

Comment:
what's `sendbuf.dtype` **before** this line?

perhaps

```
sendbuf = np.ascontiguousarray(data, dtype=self.dtype)
```

would suffice?

Reply:
No that's the issue, `sendbuf = np.ascontiguousarray(data)` is correct and has `self.dtype` as a type, the issue is that mpi4py doesn't support float16 so you have to turn it into a view of `uint16` (same size) and convert back after send receive

Reply:
What is done in the generated code too, the mpi type for float16 is USHORT then cast to Float16